src/Features/CSharp/Portable/Completion/CompletionProviders/PropertySubPatternCompletionProvider.cs

```diff
@@ -0,0 +1,177 @@
+                var typeSymbol = semanticModel.GetTypeInfo(isPattern.Expression, cancellationToken).Type;
+                if (typeSymbol != null)
+                {
+                    return (typeSymbol, token.GetLocation());
+                }
+            }
+
+            // switch (e) { case { $$
+            // switch (e) { case { P1: 0, $$
+            if (propertyPattern.IsParentKind(SyntaxKind.CasePatternSwitchLabel))
+            {
+                var casePattern = (CasePatternSwitchLabelSyntax)propertyPattern.Parent;
+                var switchStatment = casePattern.Parent?.Parent as SwitchStatementSyntax;
+                if (switchStatment != null && switchStatment.Expression != null)
+                {
+                    var typeSymbol = semanticModel.GetTypeInfo(switchStatment.Expression).Type;
+                    if (typeSymbol != null)
+                    {
+                        return (typeSymbol, token.GetLocation());
+                    }
+                }
+            }
+
+            // e is { P1: { $$
+            // e is { P1: { P2: 2, $$
+            if (propertyPattern.IsParentKind(SyntaxKind.SubpatternElement))
+            {
+                var containingProperty = ((SubpatternElementSyntax)propertyPattern.Parent).NameColon?.Name;
+                if (containingProperty != null)
+                {
+                    var symbol = semanticModel.GetSymbolInfo(containingProperty).Symbol;
+
+                    switch (symbol)
+                    {
+                        case null:
+                            break;
+                        case IPropertySymbol propertySymbol:
+                            return (propertySymbol.Type, token.GetLocation());
+                        case IFieldSymbol fieldSymbol:
+                            return (fieldSymbol.Type, token.GetLocation());
+                    }
+                }
+            }
+
+            // PROTOTYPE(NullableReferenceTypes): patterns in switch expression
+
+            return default;
+        }
+
+        // List the members that are already tested in this property sub-pattern
+        protected HashSet<string> GetTestedMembers(SyntaxTree tree, int position, CancellationToken cancellationToken)
+        {
+            var token = tree.FindTokenOnLeftOfPosition(position, cancellationToken)
+                .GetPreviousTokenIfTouchingWord(position);
+
+            // We should have gotten back a { or ,
+            if (token.Kind() == SyntaxKind.CommaToken || token.Kind() == SyntaxKind.OpenBraceToken)
+            {
+                if (token.Parent != null)
+                {
+                    if (token.Parent is PropertySubpatternSyntax subpattern)
+                    {
+                        return new HashSet<string>(subpattern.SubPatterns.Select(
+                            p => p.NameColon?.Name?.Identifier.ValueText).Where(s => !string.IsNullOrEmpty(s)));
```

Comment:
can Name be null?

Reply:
Not sure. That's why I defaulted to defensive style.

Reply:
I thought it was only constructs that are actually optional in the language (such as a name colon) that can be null

Reply:
I think you have null NameColon and Name.
No NameColon in this [example](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRIMaOQSwG4SoAOsMECAdsgDQwhTIzUAmIA1AD4ACADAARcAjAG4AsACguAZgEAmPgGE+Abwl91AmTgow+ABUEq+AcwgxhfZGYsBfNRukCALHwCyACgCUK+xo0B9PgBePhgACxxkPkijED4eGzFxPztxGyA==).

Reply:
that seems wrong. @gafter?

Reply:
I see. Property patterns use the same syntax node (SubpatternElement) as positional patterns, that's why it's optional.